lib/rubocop/ast/node_pattern/compiler.rb

```diff
@@ -0,0 +1,96 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module AST
+    class NodePattern
+      # The top-level compiler holding the global state
+      # Defers work to its subcompilers
+      #
+      # Doc on how this fits in the compiling process:
+      #   /doc/modules/ROOT/pages/node_pattern.md
+      class Compiler
+        extend Forwardable
+        attr_reader :captures, :named_parameters, :positional_parameters, :binding
+
+        def initialize
+          @temp_depth = 0 # avoid name clashes between temp variables
+          @captures = 0 # number of captures seen
+          @positional_parameters = 0 # highest % (param) number seen
+          @named_parameters = Set[] # keyword parameters
+          @binding = Binding.new # bound variables
+          @atom_subcompiler = self.class::AtomSubcompiler.new(self)
+        end
+
+        def_delegators :binding, :bind, :union_bind
+
+        def positional_parameter(number)
+          @positional_parameters = number if number > @positional_parameters
+          "param#{number}"
+        end
+
+        def named_parameter(name)
+          @named_parameters << name
+          name
+        end
+
+        def enforce_same_captures(enum)
+          return to_enum __method__, enum unless block_given?
+
+          captures_before = captures_after = nil
+          enum.each do |node|
+            captures_before ||= @captures
+            @captures = captures_before
+            yield node
+            captures_after ||= @captures
+            raise Invalid, 'each branch must have same # of captures' if captures_after != @captures
```

Comment:
Cosmetic: maybe "number" instead of "#"?

Reply:
Sure! It's the same message than currently though 😆